File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesPackage.java
package expo.modules.updates;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.facebook.react.ReactInstanceManager;
import com.facebook.react.bridge.JavaScriptContextHolder;
import com.facebook.react.bridge.ReactApplicationContext;

import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;
import expo.modules.core.BasePackage;
import expo.modules.core.ExportedModule;
import expo.modules.core.interfaces.InternalModule;
import expo.modules.core.interfaces.ReactNativeHostHandler;

// these unused imports must stay because of versioning
import expo.modules.updates.UpdatesController;

public class UpdatesPackage extends BasePackage {
  private static final String TAG = UpdatesPackage.class.getSimpleName();

  @Override
  public List<InternalModule> createInternalModules(Context context) {
    return Collections.singletonList((InternalModule) new UpdatesService(context));
  }

  @Override
  public List<ExportedModule> createExportedModules(Context context) {
    return Collections.singletonList((ExportedModule) new UpdatesModule(context));
  }

  @Override
  public List<? extends ReactNativeHostHandler> createReactNativeHostHandlers(Context context) {
    final ReactNativeHostHandler handler = new ReactNativeHostHandler() {
      private Boolean mShouldAutoSetup = null;

      @Nullable
      @Override
      public ReactInstanceManager createReactInstanceManager(boolean useDeveloperSupport) {
        if (shouldAutoSetup(context) && !useDeveloperSupport) {
          UpdatesController.initialize(context);
        }
        return null;
      }

      @Nullable
      @Override
      public String getJSBundleFile(boolean useDeveloperSupport) {
        return shouldAutoSetup(context) && !useDeveloperSupport
          ? UpdatesController.getInstance().getLaunchAssetFile()
          : null;
      }

      @Nullable
      @Override
      public String getBundleAssetName(boolean useDeveloperSupport) {
        return shouldAutoSetup(context) && !useDeveloperSupport
          ? UpdatesController.getInstance().getBundleAssetName()
          : null;
      }

      @Override
      public void onRegisterJSIModules(@NonNull ReactApplicationContext reactApplicationContext,
                                       @NonNull JavaScriptContextHolder jsContext,
                                       boolean useDeveloperSupport) {
      }

      @UiThread
      private boolean shouldAutoSetup(final Context context) {
        if (mShouldAutoSetup == null) {
          try {
            final PackageManager pm = context.getPackageManager();
            final ApplicationInfo ai = pm.getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
            mShouldAutoSetup = ai.metaData.getBoolean("expo.modules.updates.AUTO_SETUP", true);
          } catch (Exception e) {
            Log.e(TAG, "Could not read expo-updates configuration data in AndroidManifest", e);
            mShouldAutoSetup = true;
          }
        }
        return mShouldAutoSetup;
      }
    };

    return Collections.singletonList(handler);
  }
}

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesPackage.kt
package expo.modules.updates

import android.content.Context
import androidx.annotation.UiThread
import androidx.annotation.WorkerThread
import com.facebook.react.ReactActivity
import com.facebook.react.ReactNativeHost
import com.facebook.react.bridge.ReactContext
import com.facebook.react.devsupport.interfaces.DevSupportManager
import expo.modules.core.interfaces.Package
import expo.modules.core.interfaces.ReactActivityHandler
import expo.modules.core.interfaces.ReactNativeHostHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Defines the internal and exported modules for expo-updates, as well as the auto-setup behavior in
 * applicable environments.
 */
class UpdatesPackage : Package {

  override fun createReactNativeHostHandlers(context: Context): List<ReactNativeHostHandler> {
    val handler: ReactNativeHostHandler = object : ReactNativeHostHandler {

      override fun getJSBundleFile(useDeveloperSupport: Boolean): String? {
        return if (UpdatesController.instance.isActiveController) UpdatesController.instance.launchAssetFile else null
      }

      override fun getBundleAssetName(useDeveloperSupport: Boolean): String? {
        return if (UpdatesController.instance.isActiveController) UpdatesController.instance.bundleAssetName else null
      }

      override fun onWillCreateReactInstance(useDeveloperSupport: Boolean) {
        UpdatesController.initialize(context)
      }

      override fun onDidCreateDevSupportManager(devSupportManager: DevSupportManager) {
        UpdatesController.instance.onDidCreateDevSupportManager(devSupportManager)
      }

      override fun onDidCreateReactInstance(useDeveloperSupport: Boolean, reactContext: ReactContext) {
        UpdatesController.instance.onDidCreateReactInstance(reactContext)
      }

      override fun onReactInstanceException(useDeveloperSupport: Boolean, exception: Exception) {
        UpdatesController.instance.onReactInstanceException(exception)
      }
    }
    return listOf(handler)
  }

  override fun createReactActivityHandlers(activityContext: Context): List<ReactActivityHandler> {
    val handler = object : ReactActivityHandler {
      override fun getDelayLoadAppHandler(activity: ReactActivity, reactNativeHost: ReactNativeHost): ReactActivityHandler.DelayLoadAppHandler? {
        if (!BuildConfig.EX_UPDATES_ANDROID_DELAY_LOAD_APP || isUsingCustomInit) {
          return null
        }
        val context = activity.applicationContext
        val useDeveloperSupport = reactNativeHost.useDeveloperSupport
        if (!useDeveloperSupport || isUsingNativeDebug) {
          return ReactActivityHandler.DelayLoadAppHandler { whenReadyRunnable ->
            CoroutineScope(Dispatchers.IO).launch {
              startUpdatesController(context)
              invokeReadyRunnable(whenReadyRunnable)
            }
          }
        }
        return null
      }

      @WorkerThread
      private suspend fun startUpdatesController(context: Context) {
        withContext(Dispatchers.IO) {
          if (!UpdatesPackage.isUsingCustomInit) {
            UpdatesController.initialize(context)
            // Call the synchronous `launchAssetFile()` function to wait for updates ready
            UpdatesController.instance.launchAssetFile
          }
        }
      }

      @UiThread
      private suspend fun invokeReadyRunnable(whenReadyRunnable: Runnable) {
        withContext(Dispatchers.Main) {
          whenReadyRunnable.run()
        }
      }
    }

    return listOf(handler)
  }

  companion object {
    private val TAG = UpdatesPackage::class.java.simpleName
    val isUsingNativeDebug = BuildConfig.EX_UPDATES_NATIVE_DEBUG
    internal val isUsingCustomInit = BuildConfig.EX_UPDATES_CUSTOM_INIT
  }
}
